packages/next/src/server/request/params.ts

```diff
@@ -1 +1,472 @@
-export type Params = Record<string, string | string[] | undefined>
+import type { StaticGenerationStore } from '../../client/components/static-generation-async-storage.external'
+
+import { ReflectAdapter } from '../web/spec-extension/adapters/reflect'
+import {
+  abortAndThrowOnSynchronousDynamicDataAccess,
+  throwToInterruptStaticGeneration,
+  postponeWithTracking,
+} from '../app-render/dynamic-rendering'
+
+import {
+  prerenderAsyncStorage,
+  type PrerenderStore,
+} from '../app-render/prerender-async-storage.external'
+import { InvariantError } from '../../shared/lib/invariant-error'
+import { StaticGenBailoutError } from '../../client/components/static-generation-bailout'
+import { makeHangingPromise, makeResolvedReactPromise } from './utils'
+import type { FallbackRouteParams } from './fallback-params'
+
+export type Params = Record<string, string | Array<string> | undefined>
+
+/**
+ * In this version of Next.js the `params` prop passed to Layouts, Pages, and other Segments is a Promise.
+ * However to facilitate migration to this new Promise type you can currently still access params directly on the Promise instance passed to these Segments.
+ * The `UnsafeUnwrappedParams` type is available if you need to temporarily access the underlying params without first awaiting or `use`ing the Promise.
+ *
+ * In a future version of Next.js the `params` prop will be a plain Promise and this type will be removed.
+ *
+ * Typically instances of `params` can be updated automatically to be treated as a Promise by a codemod published alongside this Next.js version however if you
+ * have not yet run the codemod of the codemod cannot detect certain instances of `params` usage you should first try to refactor your code to await `params`.
+ *
+ * If refactoring is not possible but you still want to be able to access params directly without typescript errors you can cast the params Promise to this type
+ *
+ * ```tsx
+ * type Props = { params: Promise<{ id: string }>}
+ *
+ * export default async function Layout(props: Props) {
+ *  const directParams = (props.params as unknown as UnsafeUnwrappedParams<typeof props.params>)
+ *  return ...
+ * }
+ * ```
+ *
+ * This type is marked deprecated to help identify it as target for refactoring away.
+ *
+ * @deprecated
+ */
+export type UnsafeUnwrappedParams<P> =
+  P extends Promise<infer U> ? Omit<U, 'then' | 'status' | 'value'> : never
+
+export function createPrerenderParamsFromClient(
+  underlying: Params,
```

Comment:
nit: should we give these `underlying` variables a more meaningful name? maybe `underlyingParams` or just `params`?  In longer functions I found myself forgetting what `underlying` was in reference to when scanning the code.